driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,174 @@
+use super::{price_source::PriceSource, Token};
+use crate::models::TokenId;
+use anyhow::Result;
+use std::collections::HashMap;
+#[cfg(test)]
+use std::sync::mpsc::Receiver;
+use std::sync::{
+    mpsc::{self, RecvTimeoutError, Sender},
+    Arc, Mutex,
+};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+
+/// Implements `PriceSource` in a non blocking way by updating prices in a
+/// thread and reusing previous results.
+pub struct ThreadedPriceSource {
+    // Shared between this struct and the thread. The background thread writes
+    // the prices, the thread calling `get_prices` reads them.
+    price_map: Arc<Mutex<HashMap<TokenId, u128>>>,
+    // Allows the thread to notice when the owning struct is dropped.
+    _channel_to_thread: Sender<()>,
+
+    // To make testing easier we use another channel to which the thread sends
+    // a message whenever it completes one update loop.
+    #[cfg(test)]
+    test_receiver: Receiver<()>,
+}
+
+impl ThreadedPriceSource {
+    /// All token prices will be updated every `update_interval`. Prices for
+    /// other tokens will not be returned in `get_prices`.
+    ///
+    /// The join handle represents the background thread. It can be used to
+    /// verify that it exits when the created struct is dropped.
+    pub fn new<T: 'static + PriceSource + Send>(
+        tokens: Vec<Token>,
+        price_source: T,
+        update_interval: Duration,
+    ) -> (Self, JoinHandle<()>) {
+        let price_map = Arc::new(Mutex::new(HashMap::new()));
+        let (sender, receiver) = mpsc::channel::<()>();
+
+        #[cfg(test)]
+        let (test_sender, test_receiver) = mpsc::channel::<()>();
+
+        let price_map_ = price_map.clone();
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
```

Comment:
This is a super cleaver way to sleep for `update_interval` and allow waking to happen on demand! Awesome :metal:.

Reply:
In case you find it interesting the "traditional" way without channels (like in c++) would be to use a https://doc.rust-lang.org/std/sync/struct.Condvar.html which is more lightweight than a channel but a channel is much nicer to use.

Reply:
That's how I would have done it. I also find the channels way nicer to read and to use.

One minor nit, now that I think of it. Maybe the channel names an better represent what they are for - like `update_prices_{rx,tx}`. I'm not sure how others feel about this.